File: src/main/java/xylixlike/engine/Level.java
package xylixlike.engine;

import javafx.scene.Scene;
import javafx.scene.input.KeyEvent;
import javafx.scene.layout.Pane;
import javafx.scene.paint.Color;
import javafx.scene.shape.Rectangle;
import org.tinylog.Logger;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;

import static xylixlike.utils.Parsing.toDirection;

public class Level {
    boolean won;
    boolean lost;
    private List<Entity> entities;


    private Level() {
        entities = new ArrayList<>();
        won = false;
        lost = false;
    }

    public void ifOver(Runnable action) {
        if (won) {
            action.run();
        }
    }

    public void tick(KeyEvent press) {
        int movementDirection = toDirection(press.getCode());
        getEntities().forEach(actor -> actor.movementAction(movementDirection));
        collisionHandle(movementDirection);

        if (lost) {
            entities.forEach(e -> {
                e.hitbox.setTranslateX(0);
                e.hitbox.setTranslateY(0);
            });
            lost = false;
            return;
        }
    }

    private void collisionHandle(int direction) {
        int undoDirection = direction + 180;
        entities.stream().filter(e -> (e.movable)).forEach(collider -> {
            entities.stream().filter(collider::collide).forEach(collidee -> {
                String action = collider.collisionAction(collidee);
                handleAction(collider, collidee, action, undoDirection);
            });
        });
    }


    private void handleAction(Entity collider, Entity collidee, String action, int undo) {
        if (!action.isBlank()) {
            Logger.trace(action);
        }
        if ("victory".equals(action)) {
            System.out.println("You're winner!");
            won = true;
        } else if ("loss".equals(action)) {
            lost = true;
        } else if ("open".equals(action)) {
            collidee.passable = true;
            collidee.hitbox.setFill(Color.TRANSPARENT);
            collider.movable = false;
            collider.hitbox.setFill(Color.TRANSPARENT);
        } else if (!collidee.passable) {
            collider.move(undo);
        }
    }


    public List<Rectangle> getHitboxes() {
        return entities.stream().map(e -> e.hitbox).collect(Collectors.toList());
    }

    public List<Entity> getEntities() {
        return Collections.unmodifiableList(entities);
    }

    public Scene getScene() {
        Pane pane = new Pane(this.getHitboxes().toArray(Rectangle[]::new));
        Scene scene = new Scene(pane);
        scene.setOnKeyPressed(this::tick);
        return scene;
    }
}

File: src/main/kotlin/xylixlike/engine/Level.kt
package xylixlike.engine

import javafx.event.EventHandler
import javafx.scene.Scene
import javafx.scene.input.KeyEvent
import javafx.scene.layout.Pane
import javafx.scene.paint.Color
import javafx.scene.shape.Rectangle
import org.tinylog.kotlin.Logger
import xylixlike.utils.toDirection
import java.util.function.Consumer
import kotlin.streams.toList

class Level private constructor() {
    private var won: Boolean
    private var lost: Boolean
    private val entities: List<Entity>
    fun ifOver(action: Runnable) {
        if (won) {
            action.run()
        }
    }

    fun tick(press: KeyEvent) {
        val movementDirection = toDirection(press.code)
        getEntities().forEach(Consumer { actor: Entity -> actor.movementAction(movementDirection) })
        collisionHandle(movementDirection)
        if (lost) {
            entities.forEach(Consumer { e: Entity ->
                e.hitbox.translateX = 0.0
                e.hitbox.translateY = 0.0
            })
            lost = false
            return
        }
    }

    private fun collisionHandle(direction: Int) {
        val undoDirection = direction + 180
        entities.stream().filter { e: Entity -> e.movable }.forEach { collider: Entity ->
            entities.stream().filter { collidee: Entity? -> collider.collide(collidee!!) }.forEach { collidee: Entity ->
                val action = collider.collisionAction(collidee)
                handleAction(collider, collidee, action, undoDirection)
            }
        }
    }

    private fun handleAction(collider: Entity, collidee: Entity, action: String, undo: Int) {
        if (!action.isBlank()) {
            Logger.trace(action)
        }
        if ("victory" == action) {
            println("You're winner!")
            won = true
        } else if ("loss" == action) {
            lost = true
        } else if ("open" == action) {
            collidee.passable = true
            collidee.hitbox.fill = Color.TRANSPARENT
            collider.movable = false
            collider.hitbox.fill = Color.TRANSPARENT
        } else if (!collidee.passable) {
            collider.move(undo)
        }
    }

    private val hitboxes: List<Rectangle>
        get() = entities.stream().map { e: Entity -> e.hitbox }.toList()

    fun getEntities(): List<Entity> {
        return entities
    }

    val scene: Scene
        get() {
            val pane = Pane()
            pane.children.addAll(hitboxes)
            val scene = Scene(pane)
            scene.onKeyPressed = EventHandler { press: KeyEvent -> tick(press) }
            return scene
        }

    init {
        entities = ArrayList()
        won = false
        lost = false
    }
}